From ea26045ca244bd4880bf572ae128bd0446d38d91 Mon Sep 17 00:00:00 2001
From: Ananya Ghosh <ananghos@qti.qualcomm.com>
Date: Wed, 3 Dec 2025 10:46:54 +0530
Subject: [PATCH] usb: dwc3-qcom: Fix crash due to improper runtime PM handling
 during remove

The dwc3_qcom_remove() function could trigger a kernel panic because the device was allowed to enter runtime suspend while its resources were being torn down. This led to invalid memory access during late suspend callbacks. Runtime PM was not properly disabled before cleanup, allowing suspend operations to race with device removal.

- Call pm_runtime_get_sync() at the start to ensure the device is active before touching registers.
- Drop pm_runtime_allow() before disable
- Move pm_runtime_disable() to the beginning of remove() to block further PM transitions.

Change-Id: Ifd315f1d60848ec71345ac33a21b6da11835c62d
Signed-off-by: Ananya Ghosh <ananghos@qti.qualcomm.com>
---
 drivers/usb/dwc3/dwc3-qcom.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-qcom.c b/drivers/usb/dwc3/dwc3-qcom.c
index beb444576832..ae434499cc9f 100644
--- a/drivers/usb/dwc3/dwc3-qcom.c
+++ b/drivers/usb/dwc3/dwc3-qcom.c
@@ -2022,6 +2022,9 @@ static void dwc3_qcom_remove(struct platform_device *pdev)
 	struct device *dev = &pdev->dev;
 	int i, ret;
 
+	pm_runtime_get_sync(dev);
+	pm_runtime_disable(dev);
+
 	if (np) {
 		of_platform_depopulate(&pdev->dev);
 	} else {
@@ -2045,9 +2048,6 @@ static void dwc3_qcom_remove(struct platform_device *pdev)
 			  "Not able to configure phy mux selection:%d\n", ret);
 	}
 
-	pm_runtime_allow(dev);
-	pm_runtime_disable(dev);
-
 	for (i = qcom->num_clocks - 1; i >= 0; i--) {
 		if (!qcom->is_suspended)
 			clk_disable_unprepare(qcom->clks[i]);
-- 
2.34.1

